src/Badges.sol

```diff
@@ -17,4 +21,44 @@ contract Badges is ERC4973Permit {
   ) public view returns (bytes32) {
     return _getHash(from, to, tokenURI);
   }
-}
\ No newline at end of file
+
+  function getTokenIdFromHash(
+    bytes32 hash
+  ) public pure returns (uint256) {
+    return uint256(hash);
+  }
+
+
+  function mintAuthorizedBadge(
+    address from,
+    string calldata specUri,
+    bytes calldata signature
+  ) public returns (uint256) {
+    uint256 raftTokenId = _specToRaft[specUri];
+    // only registered specs can be used for minting
+    require(raftTokenId != 0, "mintAuthorizedBadge: spec is not registered");
+    // if we use this.take() it will pass in this contract's address as msg.sender
+    // so we use delegatecall to make sure we use the caller's address as msg.sender
+    (bool success, bytes memory data) = address(this).delegatecall(
+      abi.encodeWithSignature("take(address,string,bytes)", from, specUri, signature)
+    );
+    if (!success) {
+      revert("mintAuthorizedBadge: badge minting failed");
```

Comment:
Why not use the following?

```
revert(success, "mintAuthorizedBadge: badge minting failed");
```

Reply:
Revert can't take a condition like `require` ... It just takes a string. So you have to perform a check first